lib/ssoe/service.rb

```diff
@@ -56,5 +55,61 @@ def build_message(credential_method, credential_id, user, address)
         zipcode: address.zipcode
       ).perform
     end
+
+    def parse_response(response_body)
+      parsed = Hash.from_xml(response_body)
+
+      if parsed.dig('Envelope', 'Body', 'getSSOeTraitsByCSPIDResponse', 'icn')
+        icn = parsed['Envelope']['Body']['getSSOeTraitsByCSPIDResponse']['icn']
+        return { success: true, icn: } if icn.present?
+      end
+
+      if parsed.dig('Envelope', 'Body', 'Fault')
+        fault_code = parsed.dig('Envelope', 'Body', 'Fault', 'faultcode')
+        fault_string = parsed.dig('Envelope', 'Body', 'Fault', 'faultstring')
+
+        return {
+          success: false,
+          error: {
+            code: fault_code || 'UnknownError',
+            message: fault_string || 'Unable to parse SOAP response'
+          }
+        }
+      end
+
+      unknown_error
+    rescue => e
+      Rails.logger.error("[SSOe::Service::parse_response] Error parsing response: #{e.class} - #{e.message}")
+      unknown_error
+    end
+
+    def extract_icn(doc)
```

Comment:
Do you still need `extract_icn` & `extract_fault` now that you're using the Hash class XML parser and digging into that hash? I don't see these getting invoked anywhere else and they're private so it would be best to remove them if they're no longer needed.